src/gd.c

```diff
@@ -2201,9 +2201,9 @@ static void _gdImageFillTiled(gdImagePtr im, int x, int y, int nc)
 	FILL_PUSH(y+1, x, x, -1);
 	while (sp>stack) {
 		FILL_POP(y, x1, x2, dy);
-		for (x=x1; x>=0 && (!pts[y + x*wx2] && gdImageGetPixel(im,x,y)==oc); x--) {
```

Comment:
I disagree here, `x` should not be multiplied by `im->sx`. It should be multiplied by `im->sy`, this is the proper way to index the array. Since `x` is the column number, a new column starts every "height of column" pixels which is `sy` pixels. It can either be `x + y * im->sx` or `y + x * im->sy` but multiplying `x * im->sx` or `y * im->sy` would be both incorrect way to index the array and would result in out of bounds reads/writes.
Example `sx = 1000`, `sy = 1`. `pts[(100, 0)]` for `x = 100` and `y = 0`, should be `pts[100 * 1 + 0]` not `pts[100 * 100 + 0]`.

I agree it would be easier to understand if it was a 2d array, however, it would require `sy + 1` allocations which is fine but the allocation, cleanup and error recovery code would be harder to understand since a failure can occur in the middle of the allocations loop. By the way this is how the folks at PHP fixed it, but they didn't handle the case where allocation would fail during the allocations loop which could lead to memory leaks. 

So it's either more complex allocation, recovery and cleanup code or 1d array. Or may be we can have the best of both scenarios by using macros which was my initial solution but I decided against it.